### PR TITLE
[DRAFT] Task groups

### PR DESCRIPTION
**Pull request by thinkharderdev**
A somewhat hacky implementation of task groups. 

This can be made a lot more elegant but would involve a lot more annoying code changes. This is mostly intended as a way that we can test out assumptions without spending 3-4 days refactoring. 

FYI the actual task groups stuff is just the last commit. The first couple commits are just rebasing changes from upstream.

### CODE REVIEW

File: ballista/core/src/execution_plans/shuffle_writer.rs [hunk 1]
```diff
@@ -55,30 +55,7 @@ use datafusion::arrow::error::ArrowError;
...
 use datafusion::physical_plan::repartition::BatchPartitioner;
 use datafusion::physical_plan::stream::RecordBatchStreamAdapter;
-use lazy_static::lazy_static;
 use log::{debug, info};
-use lru::LruCache;
-use parking_lot::Mutex;
-use std::num::NonZeroUsize;
-use std::sync::atomic::{AtomicUsize, Ordering};
-
-lazy_static! {
```
Comment (thinkharderdev) [inline ballista/core/src/execution_plans/shuffle_writer.rs:65]: This stuff is no longer relevant once task groups are implemented so I ripped it out

File: ballista/core/src/execution_plans/coalesce_tasks.rs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.input.schema()
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        Partitioning::UnknownPartitioning(1)
```
Comment (Dandandan) [inline ballista/core/src/execution_plans/coalesce_tasks.rs:74]: One thing I couldn't find yet is how does it know how much slots to reserve for a `CoalesceTasksExec`?
(Normally that would be the number of partitions, but that's 1 now).

Comment (thinkharderdev) [reply ballista/core/src/execution_plans/coalesce_tasks.rs:74]: However many partitions are in the `MultiTaskDefinition`. So basically this is just taking the existing way in which we launch multiple tasks from the same stage in one shot but just launching them as a single `CoalesceTasksExec` (eg see `TaskManager::prepare_task_group`

Comment (thinkharderdev) [reply ballista/core/src/execution_plans/coalesce_tasks.rs:74]: That is, the rewrite to `CoalesceTasksExec` happens after the reservation is done

Comment (Dandandan) [reply ballista/core/src/execution_plans/coalesce_tasks.rs:74]: Ah yeah

**Pull request closed without merging**